### ISSUE TITLE
Maybe Konstant dice don't roll (and thus also don't change size)?

### ISSUE THREAD

**Issue by dwvanstone** [buttonmen-dev/buttonmen#1953]
I wanted to have a place to talk about these interactions:

Konstant Morphing
Konstant Weak/Mighty
Konstant Berserk

This came from discussions in #91 about Konstant Turbo suggesting that Konstant dice simply don't roll. Perhaps that could become the rule for Konstant?

If so, I'd furthermore propose that Konstant can't change size. (They are so constant they don't change at all!) This solves the problem of a N-sided die showing M for M>N. Weak/Mighty and Morphing would become no-ops. Konstant Berserk would need some thinking.

There is a brief discussion of Konstant Berserk in #1627 which discusses the Konstant die rerolling after Berserk.

**Comment by irilyth** [comment]
My two (four or five) cents about this:

(1) Remember that when a die "changes size", what actually happens is that the old die is removed from a play and a new die enters play, trigger all the things that normally happen when a die enters play. In particular, Konstant dice *do* roll when they enter play (e.g. at the start of a round); so if a Konstant die enters play in the middle of a round (via whatever means), it should roll then.

(2) The skill description says "If another skill causes a Konstant die to reroll (e.g., Chance, Trip, Ornery), it continues to show the same value." In particular, I think we've ruled before that the die *doesn't actually roll*, so it doesn't trigger things that happen when a die rolls, in which case we should probably clarify this text (e.g. "if another skill would cause a Konstant die to reroll, the die is not rerolled and the other skill has no effect"). 

(3) As a result of (2), I think that if a K+Something die would reroll, and the something skill would trigger on a reroll, then the die does not reroll and the something skill does not trigger. Mood Swing, for example, says "whenever they are rolled their size is set randomly"; since a Konstant die doesn't reroll, Mood Swing on a k(X)? die wouldn't trigger.

(4) As a result of (1), though, I think that if a K+Something die "changes size", i.e. a new die comes into play, then the new die *would* be rolled as usual. Morphing, for example, says "When a Morphing Die is used in any attack, it changes size, becoming the same size as the die that was captured. It is then re-rolled." So I think a km(6) die that is used in an attack against a (4) would be replaced by a km(4), and since a new die is entering play, it would be rolled.

Mighty and Weak are covered by (3): "When a Mighty Die rerolls for any reason, it first grows" and "When a Weak Die rerolls for any reason, it first shrinks". A Konstant die doesn't normally reroll, so Mighty or Weak on a kH(X) or kh(x) wouldn't trigger.

Berserk is covered by (4): "Once a Berserk Die performs a Berserk Attack, it is replaced with a non-berserk die". So I think a Bk(20) that makes a Berserk attack would be replaced by a k(10), and the new die would be rolled as it enters play.

Turbo is covered by (3): "After your starting roll you may change the size of a Turbo Die every time you roll it". A Konstant die doesn't normally reroll, so Turbo Swing on a k(X)! die wouldn't trigger.

What should happen if a Bk(X)! die makes a Berserk Attack? Berserk triggers, and a new k(X)! comes into play. At that point, it depends whether you think that a new die coming into play in the middle of a round is "After your starting roll", or if it *is* the starting roll for this die. I sort of thing the latter, and that Turbo wouldn't trigger.

**Comment by dwvanstone** [comment]
This is all consistent and makes sense to me. Some part of the purist in me likes konstant dice never changing in any way but your description seems clear too.

So the ambiguous cases here are Bk(X)! and mk(X)!  I think a "starting roll" would mean the first time that a die rolls (as opposed to the roll at the start of the game). And so when the new k(X)! comes into play it performs its starting roll which won't allow the user to change the size. In essence, konstant turbo dice never get to use the turbo feature.

**Comment by irilyth** [comment]
> The skill description says "If another skill causes a Konstant die to reroll (e.g., Chance, Trip, Ornery), it continues to show the same value." In particular, I think we've ruled before that the die doesn't actually roll, so it doesn't trigger things that happen when a die rolls, in which case we should probably clarify this text (e.g. "if another skill would cause a Konstant die to reroll, the die is not rerolled and the other skill has no effect").

In an IRC conversation this morning, James and I concluded that "the die doesn't actually roll" is wrong, and that Konstant dice don't reroll "after an attack" (we should maybe rephrase that as "as a result of participating in an attack", to clarify that it doesn't mean "at any time for any reason after anyone makes an attack"), but that they can reroll at other times, but they always roll the same value when they do -- but that this might trigger other skills that trigger on reroll, like Mighty or Weak.

A consequence of that ruling is that a Mighty Ornery Konstant die *would* grow and change size at the end of the turn; a Mighty Konstant Chance die could be rolled, and would grow; and a Weak or Mighty Konstant die could be the target of a Trip Attack, because it would change size, and the new size would roll when it came into play. (So a t(1) can Trip Attack a kH(10:6) or kh(10:6) because the new die might roll a 1; whereas a t(1) can't Trip Attack a k(10:6), because it won't reroll, and thus can't roll a 1.)

Anyone have strong contrary feelings about this?

**Comment by cgolubi1** [comment]
On Wed, 18 May, 2016 at 07:35:45 -0700, irilyth wrote:

> In an IRC conversation this morning, James and I concluded that "the die doesn't actually roll" is wrong, and that Konstant dice don't reroll "after an attack" (we should maybe rephrase that as "as a result of participating in an attack", to clarify that it doesn't mean "at any time for any reason after anyone makes an attack"), but that they can reroll at other times, but they always roll the same value when they do -- but that this might trigger other skills that trigger on reroll, like Mighty or Weak.

Hmm, i disagree with this.  I think the most consistent thing is for Konstant dice to be constant --- they don't reroll, and therefore they are immune to all effects which rely on things rerolling.  IMO we should update the skill help text to say this.

Logistically speaking: if what the help text says right now is really "If another skill causes a Konstant die to reroll (e.g., Chance, Trip, Ornery), it continues to show the same value." that strikes me as extremely clearly no good, because then we have to figure out how does that interact with Maximum, how does that interact with Weak?, etc.  So i don't think we can reasonably pick what the help text says right now as the correct implementation, because it's too illogical.  The only things that make sense are: (a) Konstant dice are always immune from rerolling, or (b) if Konstant dice are forced to reroll, they reroll normally.

Given those two choices, i like (a) better (and i think (a) is closer to what we have implemented right now overall --- i.e. the base case of Konstant being targeted by Trip would have to be rewritten if we went with (b)), but if other folks really want (b), i'm fine with that.  But i really don't like "Konstant dice never reroll... unless they do in which case they keep the same value... unless they can't, in which case we jump through hoops to figure it out" - that just seems way too inelegant, and IMO it's not necessary for it to be that inelegant.

**Comment by AdmiralJota** [comment]
What does the original text from the Buttons Brains set say?

**Comment by blackshadowshade** [comment]
If we decide to go with "konstant dice cannot reroll", then I would argue that they cannot be targeted by trip attacks, nor perform trip attacks. They also should not be able to perform chance rerolls.

**Comment by irilyth** [comment]
> If we decide to go with "konstant dice cannot reroll", then I would argue that they cannot be targeted by trip attacks, nor perform trip attacks. They also should not be able to perform chance rerolls.

I don't think that's necessarily a consequence: Lots of games have the idea of a targeting something with an effect that does nothing. You can cast Fireball at something that's immune to fire, it just has no effect -- the immunity to fire doesn't make you unable to try.

I'm not current with Magic: The Gathering, but at one point there was a distinction between "can't target" and "has no effect". I think "can't be the target of a Trip Attack" is very different from "the Trip Attack has no effect (the die doesn't reroll)". And I think "Konstant dice can't reroll" is perfectly compatible with the latter.

**Comment by irilyth** [comment]
(It's not actually "has no effect", because the Trip Attack might succeed, in which case it would have some effect, obviously. But the "causes the target die to reroll" effect would be countered by "this die never rerolls". That doesn't seem problematic to me.)

**Comment by irilyth** [comment]
Just to clarify my own position on this, I don't have a strong stake in either interpretation, but I do want to be consistent about it, and not have a bunch of "unless it'd be more fun to ignore that in this case" exceptions.

I'm also generally pretty willing to argue that we're *writing* the rules here, not just receiving and interpreting them. Yes, it's interesting what the original Button Brains rules said, and what the skill help text says, and what the code currently does; but if we think the game would be better if the rule were different, we should change the rule. (And note it as a house rule if we want.) My sense is that basically no one has ever agreed with me about that, so I don't bring it up very often, but I still think it. :^)

**Comment by AdmiralJota** [comment]
The difference between "can't be re-rolled" and "doesn't change its value when re-rolled" sounds pretty similar to me to the difference between "can't target" and "has no effect".

**Comment by irilyth** [comment]
> What does the original text from the Buttons Brains set say?

The last time someone asked this, Oaktree came up with this: http://www.buttonweavers.com/ui/forum.html#!threadId=50&postId=665

**Comment by irilyth** [comment]
In particular:

> After the first roll of the round, you don't re-roll the Constant Dice until the first roll of the next round. Constant Dice keep their original value for the entire round--after all, they're "constant."

That's presumably where I got the "Konstant dice never re-roll" thing from -- it's right there in the original rules.

**Comment by AdmiralJota** [comment]
I'm on the "interpreting" side of the fence whenever possible, so I'd vote in favor of "Konstant dice don't re-roll", and therefore aren't affected by Mighty or Weak.

But I'd also say that they can be tripped: nothing in the rules says they *can't* be tripped; they're just contradictory about what happens when they are. So we have to resolve the contradiction one way or the other, and saying that Konstant dice are an exception that can be tripped but don't actually re-roll when doing so seems like a better compromise than saying they're an exception that can't be tripped.

**Comment by cgolubi1** [comment]
On Wed, 18 May, 2016 at 09:09:15 -0700, AdmiralJota wrote:

> But I'd also say that they can be tripped: nothing in the rules says they *can't* be tripped; they're just contradictory about what happens when they are. So we have to resolve the contradiction one way or the other, and saying that Konstant dice are an exception that can be tripped but don't actually re-roll when doing so seems like a better compromise than saying they're an exception that can't be tripped.

I agree with this.  In particular, i think a Trip attack consists of (1) try to reroll the attacking die, (2) try to reroll the target die, (3) try to capture the target die, and that if either the attacking die or the target die doesn't reroll, that in no way precludes going on to step 3.

**Comment by blackshadowshade** [comment]
So, if we're going with the original rules from Button Brains, we also have

    A Constant Die only attacks as part of a multiple-die Skill Attack. It cannot make Power Attacks.

In that case, we'd have to say that konstant dice are basically a sub-class of stealth dice.

**Comment by blackshadowshade** [comment]
(of course, they can be targeted by other attacks, unlike stealth dice)

**Comment by blackshadowshade** [comment]
Also noting that in that above forum thread, both Julian and Oaktree interpret the rules to imply that konstant dice cannot be tripped. So I would say that there's no consensus here. If we choose to go one way or the other, it's going to be a house rule, and that's fine, just we need to be clear what we're doing.

**Comment by irilyth** [comment]
> So, if we're going with the original rules from Button Brains, we also have
>
>> A Constant Die only attacks as part of a multiple-die Skill Attack. It cannot make Power Attacks.

Right, that's a core part of how they work (and why they're so annoying :^), and how they're implemented now, right?

> In that case, we'd have to say that konstant dice are basically a sub-class of stealth dice.

I don't understand what the implication of that would be. As you note, they can be attacked by attacks other than multi-die Skill Attacks, which makes them very unlike Stealth dice. They can also make other kinds of attacks, if they have other skills: Shadow Konstant dice can make Shadow Attacks, Speed Konstant dice can make Speed Attacks, and so on.

Unless you're suggesting that we should interpret "only attacks as part of a multiple-die Skill Attack" as meaning "can not ever make any other kind of attack", and not just "cannot make Power Attacks"? That's not how either the old site or the new site interpret that rule. (The rules in Button Brains didn't address any other kinds of attacks because none of the Konstant dice in Button Brains have any other skills on them.)

> Also noting that in that above forum thread, both Julian and Oaktree interpret the rules to imply that konstant dice cannot be tripped. So I would say that there's no consensus here. If we choose to go one way or the other, it's going to be a house rule, and that's fine, just we need to be clear what we're doing.

I guess I sort of buy that the Trip Attack rules from Lunch Money are underspecified, in that they don't say what happens if you try to make a Trip Attack against a die that doesn't reroll for whatever reason. (Probably because at the time those rules were written, there were no such dice.) Given that, I agree that any ruling is a "house rule", in the sense that we have to decide *something*, and the rules as written don't say what to decide.

So, what do we have as open issues here? How close are we to coming to a final (ish :^) conclusion and forging ahead?

**Comment by blackshadowshade** [comment]
Yeah, that was my implication, that if we go strictly with the wording from Button Brains, that Konstant dice couldn't do speed, shadow, or any other type of attack apart from skill attacks.

I'm not too fussed with sticking exactly with the wording of the original rules, so I see the open issues as:
- are we unhappy enough with the status quo that we want to change it?
- are we willing to go through the coding / testing / validation process of changing the way that konstant works?

I foresee the amount of work as about:
- one week of coding the code and unit tests
- one week of adjusting the logging code
- two weeks of testing konstant intensively on the dev site
- a couple of days of wranging the correct wording of the descriptions and interactions
- the effort needed to re-educate current players

**Comment by cgolubi1** [comment]
So, just to be clear: deciding what behavior we want is very different from deciding on the prioritization of making the change.  I *do* think it's useful to decide on what behavior we want, because the next philosophical argument about skills is going to use this one as a precedent, so let's decide on what behavior we actually want.

I'm also not worried about player reeducation.  Right now this second, when a non-size-changing Konstant die is tripped, the action log says:
```
c1 performed Trip attack using [t(8):6] against [k(6):5]; Attacker t(8) rerolled 6 => 3; Defender k(6) does not reroll, was not captured. 
```

I think you're saying it's not actually fully implemented that way right now (and certainly not in the size-changing case), but players who care have been seeing that text for awhile, so if we make it so that it is fully implemented that way, i don't see the problem for players.

So, okay, with all that preamble, here is the full Konstant behavior i think we are proposing, in help text format:
* Base text (split into sentences for ease of nitpicking):
    * These dice do not reroll; after the initial roll at the beginning of each round, they keep their current value even if they are involved in attacks.
    * Konstant dice cannot perform Power attacks or single-die Skill attacks, but can perform any other attacks enabled by the skills on the die.
    * In a multi-die Skill attack, each Konstant die can add OR subtract its value to the total.
* Interactions with other skills:
    * Trip: Konstant dice can Trip and be targeted by Trip, but since they do not reroll, they will not change their value or activate any other roll-time skills during a Trip attack
    * Mighty: Mighty dice grow at roll time, and Konstant dice do not roll, so Konstant Mighty dice will not grow
    * Weak: ditto
    * Focus: Konstant Focus dice may be turned down to a new value, which they will then continue to show for the rest of the round
    * Fire: ditto
    * Chance: You may choose to reroll a Konstant Chance die, but it will be a noop because Konstant dice do not reroll

Ignoring prioritization, is that the right behavior?  If so, let's open a new issue to implement it, which can be prioritized as appropriate.

**Comment by AdmiralJota** [comment]
I mostly agree with the behavior. I think it's a little odd that we would allow you to try to "reroll" a Konstant Chance die (I'd lean towards not allowing it), but if that means less implementation work, I'm not going to argue with it.

Are you interested in comments on the phrasing, if we plan to use this directly in the help text?

**Comment by dwvanstone** [comment]
> So, okay, with all that preamble, here is the full Konstant behavior i think we are proposing, in help text format:

This is fantastic and very clear. Thank you!

**Comment by cgolubi1** [comment]
> I think it's a little odd that we would allow you to try to "reroll" a Konstant Chance die (I'd lean towards not allowing it), but if that means less implementation work, I'm not going to argue with it.

Yeah, my theory was that it would be easier to implement and would have the same gameplay effect as declining your reroll.  But i don't actually know what's easier in this case.

**Comment by cgolubi1** [comment]
> Are you interested in comments on the phrasing, if we plan to use this directly in the help text?

Hmm, let's wait for general agreement that these are the right rules first, but after that definitely.

**Comment by Hurtay** [comment]
>     * Konstant dice cannot perform Power attacks or single-die Skill attacks, but can perform any other attacks enabled by the skills on the die.
What precisely does this mean for i.e. shadow konstant dice? If you say "can perform any other attack" then a konstant die can perform a single die shadow attack? I don't think it should. Or am I wrong?

**Comment by blackshadowshade** [comment]
The point here is that there *is* no right and wrong. There's only "how we interpret the original rules" and "what we want to implement on the site".

I can see arguments both for and against allowing konstant dice to perform attacks like shadow, speed, and trip, and I continue to see no consensus.

**Comment by cgolubi1** [comment]
Just to back up a step: what we're trying to do here is *create* a working consensus, so it's only so helpful to lament that we don't have one yet.  That's what we're working on.  :>)

My two cents about Shadow Konstant - i don't think Konstant dice should be restricted from performing attacks like Speed and Trip, because those attacks are provided by a skill in addition to your base attack, so if you have that skill, you should get to perform those attacks.

However, Shadow is different --- Shadow replaces your base attack (Power) with a different base attack.  So i can see either argument --- either that Konstant dice simply can't perform Power attacks, but can perform any other type of attack that their skills allow, or that Konstant dice can't perform the non-Skill base attack (whether that's Power or Shadow).

IMO, the first is simpler (Konstant dice just can't do Power attacks), and i don't see any particular benefit to preventing Shadow.  But i'm happy to be convinced.  Hurtay, why does it seem wrong to you for a die with `ks` to perform a Shadow attack?

**Comment by AdmiralJota** [comment]
I am presently in favor of "Konstant dice just can't do Power or multi-die Skill attacks", and would permit them to do any other attacks that they would otherwise be able to perform. I don't see Shadow (adding Shadow attacks, removing Power attacks) as being that much different from, say, Berserk (adding Berserk attacks, removing Skill attacks).

**Comment by cgolubi1** [comment]
> Konstant dice just can't do Power or multi-die Skill attacks

You mean single-die skill attacks, right?

**Comment by AdmiralJota** [comment]
> You mean single-die skill attacks, right?

Yes, thanks. Edited to fix that.

**Comment by jimmosk** [comment]
Reading through this I don't see a final decision about Konstant Berserk (nor Konstant Morphing) dice, and I just had this happen in game 43137:

> jimmosk performed Berserk attack using [Bk(8):6] against [^bt(6):5,bg(8):1]; Defender ^bt(6) was captured; Defender bg(8) was captured; Attacker Bk(8) changed size from 8 to 4 sides, recipe changed from Bk(8) to k(4), does not reroll

Which says it does not reroll, but since it had been a [Bk(8):6] before, it certainly had _something_ happen to it. Did it get to that 4 via a roll, or because 6 was no longer valid so the value was set as close as possible? And should the text make explicit whatever happened?

**Comment by blackshadowshade** [comment]
From the user point of view, I agree the message is confusing.

From the dev point of view, we see a difference between "rolling into a game" and "rerolling". The newly shrunk k(4) die needs to be rolled into the game to acquire a value.

**Comment by blackshadowshade** [comment]
(and so the log message is still incorrect, I agree)
